File: zend/Excel/zend-add-in-ts/src/functions/zend/storage.ts
import { deserialize, serialize } from "./converters";
import { ZendMap, ZendType } from "./types";

export class Storage {
  dict: Map<string, ZendType>;

  constructor() {
    this.dict = new Map<string, any>();
  }

  put(name: string, value: any): void {
    this.dict.set(name, serialize(value));
  }

  putSerialized(name: string, value: ZendType): void {
    this.dict.set(name, value);
  }

  putAll(dictToPut: Map<string, any>): void {
    dictToPut.forEach((value: any, key: string) => {
      this.dict.set(key, serialize(value));
    });
  }

  putAllSerialized(dictToPut: ZendMap): void {
    if (dictToPut === undefined || dictToPut === null) return;
    for (const [key, value] of Object.entries(dictToPut)) {
      this.dict.set(key, value);
    }
  }

  get(name: string): any {
    return deserialize(this.dict.get(name));
  }

  getSerialized(name: string): ZendType {
    return this.dict.get(name);
  }

  getAll(): Map<string, any> {
    let dictCopy = new Map<string, any>();
    this.dict.forEach((value: any, key: string) => {
      dictCopy.set(key, deserialize(value));
    });

    return dictCopy;
  }

  getAllSerialized(): ZendMap {
    let dictCopy = new Map<string, ZendType>();
    this.dict.forEach((value: ZendType, key: string) => {
      dictCopy.set(key, value);
    });

    return Object.fromEntries(dictCopy.entries()) as ZendMap;
  }
}

File: zend/zend-java/src/zend/Storage.java
package zend;

import org.bson.Document;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.logging.Level;
import java.util.logging.Logger;

public class Storage {
  private Map<String, Document> dict;
  Lock lock;
  Logger log = Logger.getLogger(Storage.class.getName());

  public Storage(){
    this.dict = new HashMap<>();
    lock = new ReentrantLock();
  }

  public String getStorageAsString() {
    return new Document(dict).toJson();
  }

  public void put(String name, Document value) {
    log.log(Level.INFO, "Acquiring storage lock");
    lock.lock();
    dict.put(name, value);
    log.log(Level.INFO, "Releasing storage lock");
    lock.unlock();
  }

  public void put(String name, Object value){
    log.log(Level.INFO, "Acquiring storage lock");
    lock.lock();
    dict.put(name, Converters.serialize(value));
    log.log(Level.INFO, "Releasing storage lock");
    lock.unlock();
  }

  public Object get(String name) throws ZendException {
    log.log(Level.INFO, "Acquiring storage lock");
    lock.lock();
    Object value = null;
    if(dict.containsKey(name)){
      value = Converters.deserialize(dict.get(name));
    }
    log.log(Level.INFO, "Releasing storage lock");
    lock.unlock();
    return value;
  }

  public Document getSerialized(String name) {
    log.log(Level.INFO, "Acquiring storage lock");
    lock.lock();
    Document value = null;
    if(dict.containsKey(name)){
      value = dict.get(name);
    }
    log.log(Level.INFO, "Releasing storage lock");
    lock.unlock();
    return value;
  }

  public Map<String, Document> getAllSerialized(){
    log.log(Level.INFO, "Acquiring storage lock");
    lock.lock();
    Map<String, Document> dictCopy = new HashMap<>();
    for (Map.Entry<String, Document> entry : dict.entrySet()) {
      dictCopy.put(entry.getKey(), entry.getValue());
    }
    log.log(Level.INFO, "Releasing storage lock");
    lock.unlock();
    return dictCopy;
  }

  public Map<String, Object> getAll() throws ZendException {
    log.log(Level.INFO, "Acquiring storage lock");
    lock.lock();
    Map<String, Object> dictCopy = new HashMap<>();
    for (Map.Entry<String, Document> entry : dict.entrySet()) {
      dictCopy.put(entry.getKey(), Converters.deserialize(entry.getValue()));
    }
    log.log(Level.INFO, "Releasing storage lock");
    lock.unlock();
    return dictCopy;
  }

  public void putAllFromSerializedMap(Map<String, Document> toAdd){
    log.log(Level.INFO, "Acquiring storage lock");
    lock.lock();
    for (Map.Entry<String, Document> entry : toAdd.entrySet()) {
      dict.put(entry.getKey(), entry.getValue());
    }
    log.log(Level.INFO, "Releasing storage lock");
    lock.unlock();
  }

  public void putAllFromMap(Map<String, Object> toAdd){
    log.log(Level.INFO, "Acquiring storage lock");
    lock.lock();
    for (Map.Entry<String, Object> entry : toAdd.entrySet()) {
      dict.put(entry.getKey(), Converters.serialize(entry.getValue()));
    }
    log.log(Level.INFO, "Releasing storage lock");
    lock.unlock();
  }
}